File: apps/hepta-native/src/shell-runtime.js
const STABLE_ID = /^[A-Za-z0-9._:-]{1,128}$/;
const DIGEST = /^[0-9a-f]{64}$/;
const ZERO_DIGEST = "0".repeat(64);
const PLATFORM_ACTIONS = new Set([
  "open_path",
  "reveal_path",
  "copy_text",
  "notify",
]);

function record(value, name) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new TypeError(`${name} must be an object`);
  }
  return value;
}

function stableId(value, name) {
  if (typeof value !== "string" || !STABLE_ID.test(value)) {
    throw new TypeError(`${name} must be a bounded stable identifier`);
  }
  return value;
}

function digest(value, name) {
  if (typeof value !== "string" || !DIGEST.test(value) || value === ZERO_DIGEST) {
    throw new TypeError(`${name} must be a non-zero lowercase SHA-256 digest`);
  }
  return value;
}

function positive(value, name) {
  if (!Number.isSafeInteger(value) || value < 1) {
    throw new TypeError(`${name} must be a positive safe integer`);
  }
  return value;
}

function result(value) {
  return Object.freeze({
    ...value,
    authorityGranted: false,
    filesystemAuthority: false,
    notificationAuthority: false,
    updateAuthority: false,
  });
}

export class NativeShellRuntime {
  #backend;
  #platform;
  #updater;
  #session = null;
  #view = null;
  #operations = new Map();

  constructor({ backend, platform, updater }) {
    for (const [name, value, methods] of [
      ["backend", backend, ["connect", "request", "close"]],
      ["platform", platform, ["permission", "invoke"]],
      ["updater", updater, ["verify", "apply", "rollback"]],
    ]) {
      record(value, name);
      for (const method of methods) {
        if (typeof value[method] !== "function") {
          throw new TypeError(`${name}.${method} must be a function`);
        }
      }
    }
    this.#backend = backend;
    this.#platform = platform;
    this.#updater = updater;
  }

  async connectRuntime(manifest) {
    record(manifest, "manifest");
    const endpointId = stableId(manifest.endpointId, "endpointId");
    const manifestDigest = digest(manifest.manifestDigest, "manifestDigest");
    const protocolVersion = positive(manifest.protocolVersion, "protocolVersion");
    const observed = record(
      await this.#backend.connect({ endpointId, manifestDigest, protocolVersion }),
      "backend connection",
    );
    if (observed.authenticated !== true) {
      throw new TypeError("backend connection is not authenticated");
    }
    if (observed.protocolVersion !== protocolVersion) {
      throw new TypeError("backend protocol version mismatch");
    }
    this.#session = {
      endpointId,
      manifestDigest,
      protocolVersion,
      sessionId: stableId(observed.sessionId, "sessionId"),
      generation: positive(observed.generation, "generation"),
    };
    this.#view = null;
    return result({ kind: "NativeSessionV1", ...this.#session });
  }

  renderRuntimeView(view) {
    this.#requireSession();
    record(view, "view");
    if (view.sessionId !== this.#session.sessionId) {
      throw new TypeError("view session mismatch");
    }
    if (view.sessionGeneration !== this.#session.generation) {
      throw new TypeError("view session generation mismatch");
    }
    const generation = positive(view.generation, "view generation");
    const revision = positive(view.revision, "view revision");
    const viewDigest = digest(view.digest, "view digest");
    if (this.#view) {
      if (generation < this.#view.generation) {
        throw new TypeError("view generation regressed");
      }
      if (generation === this.#view.generation && revision <= this.#view.revision) {
        throw new TypeError("view revision did not advance");
      }
    }
    this.#view = {
      generation,
      revision,
      digest: viewDigest,
      modules: Object.freeze([...(view.modules ?? [])]),
    };
    return result({
      kind: "NativePresentationStateV1",
      sessionId: this.#session.sessionId,
      ...this.#view,
      stale: false,
    });
  }

  async requestPlatformCapability(input) {
    this.#requireView();
    record(input, "input");
    const operationId = stableId(input.operationId, "operationId");
    if (!PLATFORM_ACTIONS.has(input.action)) {
      throw new TypeError("platform action is not registered");
    }
    if (input.displayedRevision !== this.#view.revision) {
      throw new TypeError("platform request was confirmed against a stale view");
    }
    const finalPayloadDigest = digest(input.finalPayloadDigest, "finalPayloadDigest");
    const grantPayloadDigest = digest(input.grantPayloadDigest, "grantPayloadDigest");
    if (finalPayloadDigest !== grantPayloadDigest) {
      throw new TypeError("grant does not bind final platform payload");
    }
    const prior = this.#operations.get(operationId);
    if (prior) {
      if (prior.finalPayloadDigest !== finalPayloadDigest) {
        throw new TypeError("operation identity was reused with changed payload");
      }
      return prior.receipt;
    }
    const permission = record(
      await this.#platform.permission({
        action: input.action,
        resource: input.resource,
      }),
      "platform permission",
    );
    if (permission.allowed !== true) {
      return result({
        kind: "PlatformDecisionV1",
        operationId,
        action: input.action,
        status: "rejected",
        terminalObserved: true,
        outcomeDigest: digest(permission.outcomeDigest, "outcomeDigest"),
      });
    }
    const observed = record(
      await this.#platform.invoke({
        sessionId: this.#session.sessionId,
        sessionGeneration: this.#session.generation,
        operationId,
        action: input.action,
        resource: input.resource,
        finalPayloadDigest,
      }),
      "platform observation",
    );
    let receipt;
    if (observed.terminalObserved !== true) {
      receipt = result({
        kind: "PlatformDecisionV1",
        operationId,
        action: input.action,
        status: "indeterminate",
        terminalObserved: false,
        outcomeDigest: null,
      });
    } else {
      if (observed.status !== "succeeded" && observed.status !== "failed") {
        throw new TypeError("terminal platform status is not registered");
      }
      receipt = result({
        kind: "PlatformDecisionV1",
        operationId,
        action: input.action,
        status: observed.status,
        terminalObserved: true,
        outcomeDigest: digest(observed.outcomeDigest, "outcomeDigest"),
      });
    }
    this.#operations.set(operationId, { finalPayloadDigest, receipt });
    return receipt;
  }

  async applyShellUpdate(input) {
    this.#requireSession();
    record(input, "input");
    const packageDigest = digest(input.packageDigest, "packageDigest");
    const predecessorDigest = digest(input.predecessorDigest, "predecessorDigest");
    const evidenceDigest = digest(input.evidenceDigest, "evidenceDigest");
    const selectedBy = stableId(input.selectedBy, "selectedBy");
    if (selectedBy === input.generatorPrincipal) {
      throw new TypeError("shell update cannot be selected by its generator");
    }
    const verification = record(
      await this.#updater.verify({
        packageDigest,
        predecessorDigest,
        evidenceDigest,
        platform: input.platform,
        architecture: input.architecture,
        backendProtocolVersion: this.#session.protocolVersion,
      }),
      "update verification",
    );
    if (verification.accepted !== true) {
      throw new TypeError("shell update verification failed");
    }
    const observed = record(
      await this.#updater.apply({
        packageDigest,
        predecessorDigest,
        restartRequired: true,
      }),
      "update observation",
    );
    if (observed.terminalObserved !== true || observed.restarted !== true) {
      await this.#updater.rollback({ predecessorDigest });
      return result({
        kind: "UpdateDispositionV1",
        packageDigest,
        predecessorDigest,
        status: "quarantined",
        terminalObserved: observed.terminalObserved === true,
      });
    }
    return result({
      kind: "UpdateDispositionV1",
      packageDigest,
      predecessorDigest,
      status: "succeeded",
      terminalObserved: true,
    });
  }

  async close() {
    if (!this.#session) {
      return;
    }
    await this.#backend.close({ sessionId: this.#session.sessionId });
    this.#session = null;
    this.#view = null;
  }

  #requireSession() {
    if (!this.#session) {
      throw new TypeError("native shell is not connected");
    }
  }

  #requireView() {
    this.#requireSession();
    if (!this.#view) {
      throw new TypeError("platform request requires a coherent runtime view");
    }
  }
}

File: codex-rs/hepta-native-gateway/src/shell_runtime.rs
use std::collections::HashMap;
use std::hash::Hash;

use anyhow::Context;
use anyhow::Result;
use anyhow::bail;

const MAX_RESOURCE_BYTES: usize = 4096;
const MAX_MODULES: usize = 512;
const ZERO_DIGEST: &str = "0000000000000000000000000000000000000000000000000000000000000000";

#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub(crate) enum PlatformAction {
    OpenPath,
    RevealPath,
    CopyText,
    Notify,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub(crate) struct SessionKey {
    pub(crate) session_id: String,
    pub(crate) generation: u64,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct EndpointManifest {
    pub(crate) endpoint_id: String,
    pub(crate) manifest_digest: String,
    pub(crate) protocol_version: u64,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct BackendSession {
    pub(crate) authenticated: bool,
    pub(crate) protocol_version: u64,
    pub(crate) session_id: String,
    pub(crate) generation: u64,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct ViewInput {
    pub(crate) session: SessionKey,
    pub(crate) generation: u64,
    pub(crate) revision: u64,
    pub(crate) digest: String,
    pub(crate) modules: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct ViewState {
    pub(crate) generation: u64,
    pub(crate) revision: u64,
    pub(crate) digest: String,
    pub(crate) modules: Vec<String>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct PlatformRequest {
    pub(crate) operation_id: String,
    pub(crate) action: PlatformAction,
    pub(crate) resource: String,
    pub(crate) displayed_revision: u64,
    pub(crate) payload_digest: String,
    pub(crate) grant: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct VerifiedPlatformGrant {
    pub(crate) session: SessionKey,
    pub(crate) operation_id: String,
    pub(crate) action: PlatformAction,
    pub(crate) payload_digest: String,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum TerminalStatus {
    Succeeded,
    Failed,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum PermissionDecision {
    Allowed,
    Denied { outcome_digest: String },
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum ReconcileObservation {
    NotFound,
    Indeterminate,
    Terminal {
        status: TerminalStatus,
        outcome_digest: String,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) enum InvokeObservation {
    Indeterminate,
    Terminal {
        status: TerminalStatus,
        outcome_digest: String,
    },
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum PlatformStatus {
    Rejected,
    Indeterminate,
    Succeeded,
    Failed,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub(crate) struct PlatformReceipt {
    pub(crate) session: SessionKey,
    pub(crate) operation_id: String,
    pub(crate) action: PlatformAction,
    pub(crate) status: PlatformStatus,
    pub(crate) terminal_observed: bool,
    pub(crate) outcome_digest: Option<String>,
}

pub(crate) trait BackendConnector: Send + Sync {
    fn connect(&self, manifest: &EndpointManifest) -> Result<BackendSession>;
    fn close(&self, session: &SessionKey) -> Result<()>;
}

pub(crate) trait PlatformAdapter: Send + Sync {
    fn permission(&self, request: &PlatformRequest) -> Result<PermissionDecision>;
    fn reconcile(
        &self,
        session: &SessionKey,
        request: &PlatformRequest,
    ) -> Result<ReconcileObservation>;
    fn invoke(&self, session: &SessionKey, request: &PlatformRequest) -> Result<InvokeObservation>;
}

pub(crate) trait GrantVerifier: Send + Sync {
    fn verify(&self, session: &SessionKey, request: &PlatformRequest)
    -> Result<VerifiedPlatformGrant>;
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
struct OperationKey {
    session: SessionKey,
    operation_id: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct OperationRecord {
    payload_digest: String,
    receipt: PlatformReceipt,
}

pub(crate) struct NativeShellRuntime<B, P, G> {
    backend: B,
    platform: P,
    grants: G,
    session: Option<SessionKey>,
    view: Option<ViewState>,
    operations: HashMap<OperationKey, OperationRecord>,
}

impl<B, P, G> NativeShellRuntime<B, P, G>
where
    B: BackendConnector,
    P: PlatformAdapter,
    G: GrantVerifier,
{
    pub(crate) fn new(backend: B, platform: P, grants: G) -> Self {
        Self {
            backend,
            platform,
            grants,
            session: None,
            view: None,
            operations: HashMap::new(),
        }
    }

    pub(crate) fn connect_runtime(&mut self, manifest: EndpointManifest) -> Result<SessionKey> {
        validate_stable_id(&manifest.endpoint_id, "endpoint_id")?;
        validate_digest(&manifest.manifest_digest, "manifest_digest")?;
        validate_positive(manifest.protocol_version, "protocol_version")?;
        let observed = self.backend.connect(&manifest)?;
        if !observed.authenticated {
            bail!("backend connection is not authenticated");
        }
        if observed.protocol_version != manifest.protocol_version {
            bail!("backend protocol version mismatch");
        }
        validate_stable_id(&observed.session_id, "session_id")?;
        validate_positive(observed.generation, "session generation")?;

        let session = SessionKey {
            session_id: observed.session_id,
            generation: observed.generation,
        };
        self.session = Some(session.clone());
        self.view = None;
        self.operations.clear();
        Ok(session)
    }

    pub(crate) fn render_runtime_view(&mut self, input: ViewInput) -> Result<ViewState> {
        let session = self.require_session()?;
        if &input.session != session {
            bail!("view session mismatch");
        }
        validate_positive(input.generation, "view generation")?;
        validate_positive(input.revision, "view revision")?;
        validate_digest(&input.digest, "view digest")?;
        if input.modules.len() > MAX_MODULES {
            bail!("view module count exceeds {MAX_MODULES}");
        }
        for module in &input.modules {
            validate_stable_id(module, "module id")?;
        }
        if let Some(previous) = &self.view {
            if input.generation < previous.generation {
                bail!("view generation regressed");
            }
            if input.generation == previous.generation && input.revision <= previous.revision {
                bail!("view revision did not advance");
            }
        }
        let view = ViewState {
            generation: input.generation,
            revision: input.revision,
            digest: input.digest,
            modules: input.modules,
        };
        self.view = Some(view.clone());
        Ok(view)
    }

    pub(crate) fn request_platform_capability(
        &mut self,
        request: PlatformRequest,
    ) -> Result<PlatformReceipt> {
        let session = self.require_session()?.clone();
        let view = self.view.as_ref().context("platform request requires a coherent runtime view")?;
        validate_stable_id(&request.operation_id, "operation_id")?;
        validate_bounded_text(&request.resource, "resource")?;
        validate_digest(&request.payload_digest, "payload_digest")?;
        if request.displayed_revision != view.revision {
            bail!("platform request was confirmed against a stale view");
        }
        let grant = self.grants.verify(&session, &request)?;
        if grant.session != session
            || grant.operation_id != request.operation_id
            || grant.action != request.action
            || grant.payload_digest != request.payload_digest
        {
            bail!("verified grant does not bind the final platform request");
        }

        let key = OperationKey {
            session: session.clone(),
            operation_id: request.operation_id.clone(),
        };
        if let Some(prior) = self.operations.get(&key) {
            if prior.payload_digest != request.payload_digest {
                bail!("operation identity was reused with changed payload");
            }
            if prior.receipt.terminal_observed {
                return Ok(prior.receipt.clone());
            }
            return self.reconcile_indeterminate(key, request);
        }

        match self.platform.reconcile(&session, &request)? {
            ReconcileObservation::NotFound => {}
            ReconcileObservation::Indeterminate => {
                let receipt = indeterminate_receipt(&session, &request);
                self.remember(key, &request, receipt.clone());
                return Ok(receipt);
            }
            ReconcileObservation::Terminal {
                status,
                outcome_digest,
            } => {
                validate_digest(&outcome_digest, "outcome_digest")?;
                let receipt = terminal_receipt(&session, &request, status, outcome_digest);
                self.remember(key, &request, receipt.clone());
                return Ok(receipt);
            }
        }

        match self.platform.permission(&request)? {
            PermissionDecision::Denied { outcome_digest } => {
                validate_digest(&outcome_digest, "outcome_digest")?;
                let receipt = PlatformReceipt {
                    session,
                    operation_id: request.operation_id.clone(),
                    action: request.action,
                    status: PlatformStatus::Rejected,
                    terminal_observed: true,
                    outcome_digest: Some(outcome_digest),
                };
                self.remember(key, &request, receipt.clone());
                Ok(receipt)
            }
            PermissionDecision::Allowed => {
                let receipt = match self.platform.invoke(&session, &request)? {
                    InvokeObservation::Indeterminate => indeterminate_receipt(&session, &request),
                    InvokeObservation::Terminal {
                        status,
                        outcome_digest,
                    } => {
                        validate_digest(&outcome_digest, "outcome_digest")?;
                        terminal_receipt(&session, &request, status, outcome_digest)
                    }
                };
                self.remember(key, &request, receipt.clone());
                Ok(receipt)
            }
        }
    }

    pub(crate) fn close(&mut self) -> Result<()> {
        if let Some(session) = self.session.take() {
            self.backend.close(&session)?;
        }
        self.view = None;
        self.operations.clear();
        Ok(())
    }

    fn reconcile_indeterminate(
        &mut self,
        key: OperationKey,
        request: PlatformRequest,
    ) -> Result<PlatformReceipt> {
        let session = key.session.clone();
        let receipt = match self.platform.reconcile(&session, &request)? {
            ReconcileObservation::Terminal {
                status,
                outcome_digest,
            } => {
                validate_digest(&outcome_digest, "outcome_digest")?;
                terminal_receipt(&session, &request, status, outcome_digest)
            }
            ReconcileObservation::NotFound | ReconcileObservation::Indeterminate => {
                indeterminate_receipt(&session, &request)
            }
        };
        self.remember(key, &request, receipt.clone());
        Ok(receipt)
    }

    fn remember(&mut self, key: OperationKey, request: &PlatformRequest, receipt: PlatformReceipt) {
        self.operations.insert(
            key,
            OperationRecord {
                payload_digest: request.payload_digest.clone(),
                receipt,
            },
        );
    }

    fn require_session(&self) -> Result<&SessionKey> {
        self.session.as_ref().context("native shell is not connected")
    }
}

fn terminal_receipt(
    session: &SessionKey,
    request: &PlatformRequest,
    status: TerminalStatus,
    outcome_digest: String,
) -> PlatformReceipt {
    PlatformReceipt {
        session: session.clone(),
        operation_id: request.operation_id.clone(),
        action: request.action,
        status: match status {
            TerminalStatus::Succeeded => PlatformStatus::Succeeded,
            TerminalStatus::Failed => PlatformStatus::Failed,
        },
        terminal_observed: true,
        outcome_digest: Some(outcome_digest),
    }
}

fn indeterminate_receipt(session: &SessionKey, request: &PlatformRequest) -> PlatformReceipt {
    PlatformReceipt {
        session: session.clone(),
        operation_id: request.operation_id.clone(),
        action: request.action,
        status: PlatformStatus::Indeterminate,
        terminal_observed: false,
        outcome_digest: None,
    }
}

fn validate_positive(value: u64, name: &str) -> Result<()> {
    if value == 0 {
        bail!("{name} must be positive");
    }
    Ok(())
}

fn validate_stable_id(value: &str, name: &str) -> Result<()> {
    if value.is_empty()
        || value.len() > 128
        || !value
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b':' | b'-'))
    {
        bail!("{name} must be a bounded stable identifier");
    }
    Ok(())
}

fn validate_digest(value: &str, name: &str) -> Result<()> {
    if value.len() != 64
        || value == ZERO_DIGEST
        || !value.bytes().all(|byte| byte.is_ascii_digit() || matches!(byte, b'a'..=b'f'))
    {
        bail!("{name} must be a non-zero lowercase SHA-256 digest");
    }
    Ok(())
}

fn validate_bounded_text(value: &str, name: &str) -> Result<()> {
    if value.is_empty() || value.len() > MAX_RESOURCE_BYTES || value.contains('\0') {
        bail!("{name} must be non-empty, NUL-free and at most {MAX_RESOURCE_BYTES} bytes");
    }
    Ok(())
}
